File: mcp-server/internal/domains/skills/fetch.go
package skills

import (
	"context"
	"fmt"

	"github.com/stigmer/stigmer/mcp-server/internal/domains"
	skillv1 "github.com/stigmer/stigmer/mcp-server/proto/ai/stigmer/agentic/skill/v1"
	"github.com/stigmer/stigmer/mcp-server/proto/ai/stigmer/commons/apiresource"
	"github.com/stigmer/stigmer/mcp-server/proto/ai/stigmer/commons/apiresource/apiresourcekind"
	"google.golang.org/grpc"
)

// Fetch retrieves a skill by org, slug, and optional version, returning its
// JSON representation. Pass an empty version string to get the latest version.
// Both tool and resource handlers delegate to this function for the actual RPC.
func Fetch(ctx context.Context, serverAddress, org, slug, version string) (string, error) {
	return domains.WithConnection(ctx, serverAddress,
		func(ctx context.Context, conn *grpc.ClientConn) (string, error) {
			client := skillv1.NewSkillQueryControllerClient(conn)
			skill, err := client.GetByReference(ctx, &apiresource.ApiResourceReference{
				Org:     org,
				Kind:    apiresourcekind.ApiResourceKind_skill,
				Slug:    slug,
				Version: version,
			})
			if err != nil {
				return "", domains.RPCError(err, fmt.Sprintf("skill %q in org %q", slug, org))
			}
			return domains.MarshalJSON(skill)
		})
}

File: mcp-server-ts/src/domains/skills/fetch.ts
// Skill read path: the single RPC the get_skill tool and both skill resource
// templates (latest + versioned) delegate to.
// Go parity: mcp-server/internal/domains/skills/fetch.go.
//
// Skills are versioned: an empty version string requests the latest version; a
// tag name or SHA-256 hash requests a specific one.

import { SkillSchema } from "@stigmer/protos/ai/stigmer/agentic/skill/v1/api_pb";
import { SkillQueryController } from "@stigmer/protos/ai/stigmer/agentic/skill/v1/query_pb";
import { ApiResourceKind } from "@stigmer/protos/ai/stigmer/commons/apiresource/apiresourcekind/api_resource_kind_pb";

import { withClient } from "../client";
import { toProtoJson } from "../marshal";
import { rpcError } from "../rpcerr";

/**
 * Retrieve a skill by org, slug, and optional version, returning its protojson
 * representation. Pass an empty version string for the latest version.
 */
export async function fetchSkill(
  serverAddress: string,
  token: string,
  org: string,
  slug: string,
  version: string,
): Promise<string> {
  return withClient(SkillQueryController, serverAddress, token, async (client, callOptions) => {
    try {
      const skill = await client.getByReference(
        { org, kind: ApiResourceKind.skill, slug, version },
        callOptions,
      );
      return toProtoJson(SkillSchema, skill);
    } catch (err) {
      throw rpcError(err, `skill "${slug}" in org "${org}"`);
    }
  });
}
